File: src/main/java/ch/uzh/ifi/access/repository/AssignmentRepository.java
package ch.uzh.ifi.access.repository;

import ch.uzh.ifi.access.model.Assignment;
import ch.uzh.ifi.access.projections.AssignmentWorkspace;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.security.access.prepost.PostAuthorize;
import org.springframework.security.access.prepost.PostFilter;

import java.util.List;
import java.util.Optional;

public interface AssignmentRepository extends JpaRepository<Assignment, Long> {

    @PostFilter("hasRole(#courseSlug + '-assistant') or (hasRole(#courseSlug) and filterObject.published)")
    List<AssignmentWorkspace> findByCourse_SlugOrderByOrdinalNumDesc(String courseSlug);

    @PostAuthorize("hasRole(#courseSlug + '-assistant') or (hasRole(#courseSlug) and returnObject.present and returnObject.get().published)")
    Optional<AssignmentWorkspace> findByCourse_SlugAndSlug(String courseSlug, String assignmentSlug);

    Optional<Assignment> getByCourse_SlugAndSlug(String courseSlug, String assignmentSlug);

    boolean existsByCourse_SlugAndSlug(String courseSlug, String assignmentSlug);

}

File: src/main/kotlin/ch/uzh/ifi/access/repository/AssignmentRepository.kt
package ch.uzh.ifi.access.repository

import ch.uzh.ifi.access.model.Assignment
import ch.uzh.ifi.access.projections.AssignmentWorkspace
import ch.uzh.ifi.access.projections.AssignmentCourseView
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.security.access.prepost.PostAuthorize
import org.springframework.security.access.prepost.PostFilter

interface AssignmentRepository : JpaRepository<Assignment?, Long?> {
    @PostFilter("filterObject.enabled and (hasRole(#courseSlug + '-assistant') or (hasRole(#courseSlug) and filterObject.isPublished) or hasRole(#courseSlug + '-supervisor'))")
    fun findByCourse_SlugOrderByOrdinalNumDesc(courseSlug: String?): List<AssignmentWorkspace>

    // Course-page variant: same visibility rule as findByCourse_SlugOrderByOrdinalNumDesc, lighter projection (no task list).
    @PostFilter("filterObject.enabled and (hasRole(#courseSlug + '-assistant') or (hasRole(#courseSlug) and filterObject.isPublished) or hasRole(#courseSlug + '-supervisor'))")
    fun findCourseViewsByCourse_SlugOrderByOrdinalNumDesc(courseSlug: String?): List<AssignmentCourseView>

    // TODO: implement visibility based on date
    //@PostAuthorize("hasRole(#courseSlug + '-assistant') or (hasRole(#courseSlug) and returnObject.present and returnObject.get().published)")
    @PostAuthorize("hasRole(#courseSlug + '-assistant') or (hasRole(#courseSlug))") // and returnObject.present)")
    fun findByCourse_SlugAndSlug(courseSlug: String?, assignmentSlug: String?): AssignmentWorkspace?
    fun getByCourse_SlugAndSlug(courseSlug: String?, assignmentSlug: String?): Assignment?
    fun existsByCourse_SlugAndSlug(courseSlug: String?, assignmentSlug: String?): Boolean
}
